vmdb/spec/models/ems_openstack_infra_spec.rb

```diff
@@ -44,4 +44,36 @@
     end
   end
 
+  context "provider_hooks" do
+    before :each do
+      @ems = FactoryGirl.create(:ems_openstack_infra_with_authentication)
+    end
+
+    it "creates related ProviderOpenstack after creating EmsOpenstackInfra" do
+      @ems.provider.name.should equal @ems.name
+      @ems.provider.zone.should == @ems.zone
+      ProviderOpenstack.count.should equal 1
+    end
+
+    it "destroys related ProviderOpenstack after destroying EmsOpenstackInfra" do
+      ProviderOpenstack.count.should equal 1
+      @ems.destroy
+      ProviderOpenstack.count.should equal 0
+    end
+
+    it "related EmsOpenstack nullifies relation to ProviderOpenstack on EmsOpenstackInfra destroy" do
+      # add ems_cloud relation to @ems
+      @ems_cloud = FactoryGirl.create(:ems_openstack_with_authentication)
+      @ems.provider.cloud_ems << @ems_cloud
```

Comment:
It's been a while but doesn't the `<<` shovel operator on a has_many automatically save the `provider` in this instance?  Is the next line needed?